File: code/causal_inference/regression_discontinuity.R
# Regression Discontinuity Design (RDD) in R
# ===========================================
#
# Implementation using rdrobust package for robust RDD inference
#
# References:
#   Imbens & Lemieux (2008), Calonico et al. (2014, 2020)

# Load required packages
suppressPackageStartupMessages({
  library(rdrobust)  # For robust RDD inference
  library(rddensity) # For density continuity tests
  library(ggplot2)
  library(dplyr)
  library(tidyr)
})

# ============================================
# Simulate RDD Data
# ============================================

simulate_rdd_data <- function(n = 1000, effect = 5.0, cutoff = 0,
                             polynomial_order = 2, seed = 42) {
  #' Simulate sharp RDD data
  #'
  #' @param n Sample size
  #' @param effect True treatment effect
  #' @param cutoff Cutoff for treatment assignment
  #' @param polynomial_order Order of baseline polynomial
  #' @return data.frame with running variable, outcome, treatment

  set.seed(seed)

  # Running variable
  x <- runif(n, -2, 2)

  # Treatment assignment (sharp RDD)
  D <- as.integer(x >= cutoff)

  # Outcome with polynomial baseline
  y_baseline <- 10 + 2*x + 0.5*x^2

  if (polynomial_order >= 3) {
    y_baseline <- y_baseline + 0.1*x^3
  }

  # Add treatment effect
  y <- y_baseline + effect * D + rnorm(n, 0, 1)

  data.frame(
    running_var = x,
    outcome = y,
    treatment = D
  )
}

# ============================================
# RDD Estimation using rdrobust
# ============================================

estimate_rdd <- function(data, running_var = "running_var",
                        outcome = "outcome", cutoff = 0,
                        bandwidth = NULL, kernel = "triangular") {
  #' Estimate RDD using rdrobust
  #'
  #' @param data data.frame
  #' @param running_var Name of running variable column
  #' @param outcome Name of outcome column
  #' @param cutoff Cutoff value
  #' @param bandwidth Bandwidth (NULL for optimal)
  #' @param kernel Kernel type
  #' @return rdrobust object

  x <- data[[running_var]]
  y <- data[[outcome]]

  # Estimate RDD with robust inference
  if (is.null(bandwidth)) {
    # Use MSE-optimal bandwidth
    rdd_result <- rdrobust(y, x, c = cutoff, kernel = kernel)
  } else {
    rdd_result <- rdrobust(y, x, c = cutoff, h = bandwidth, kernel = kernel)
  }

  rdd_result
}

# ============================================
# RDD Visualization
# ============================================

plot_rdd <- function(data, rdd_result, running_var = "running_var",
                    outcome = "outcome", cutoff = 0, n_bins = 20) {
  #' Create RDD visualization with binned scatter plot
  #'
  #' @param data data.frame
  #' @param rdd_result rdrobust object
  #' @param running_var Running variable name
  #' @param outcome Outcome variable name
  #' @param cutoff Cutoff value
  #' @param n_bins Number of bins

  x <- data[[running_var]]
  y <- data[[outcome]]

  # Create bins
  x_left <- x[x < cutoff]
  x_right <- x[x >= cutoff]
  y_left <- y[x < cutoff]
  y_right <- y[x >= cutoff]

  # Bin means
  bins_left <- cut(x_left, breaks = n_bins/2)
  bins_right <- cut(x_right, breaks = n_bins/2)

  binned_left <- data.frame(x = x_left, y = y_left, bin = bins_left) %>%
    group_by(bin) %>%
    summarise(x_mean = mean(x), y_mean = mean(y), .groups = "drop")

  binned_right <- data.frame(x = x_right, y = y_right, bin = bins_right) %>%
    group_by(bin) %>%
    summarise(x_mean = mean(x), y_mean = mean(y), .groups = "drop")

  # Get bandwidth
  h <- rdd_result$bws[1]

  # Fit polynomials within bandwidth
  mask_left <- (x < cutoff) & (abs(x - cutoff) <= h)
  mask_right <- (x >= cutoff) & (abs(x - cutoff) <= h)

  if (sum(mask_left) > 0) {
    fit_left <- lm(y[mask_left] ~ x[mask_left])
    x_plot_left <- seq(min(x[mask_left]), cutoff, length.out = 100)
    y_plot_left <- predict(fit_left, newdata = data.frame(x = x_plot_left))
  }

  if (sum(mask_right) > 0) {
    fit_right <- lm(y[mask_right] ~ x[mask_right])
    x_plot_right <- seq(cutoff, max(x[mask_right]), length.out = 100)
    y_plot_right <- predict(fit_right, newdata = data.frame(x = x_plot_right))
  }

  # Create plot
  p <- ggplot() +
    # Binned scatter
    geom_point(data = binned_left, aes(x = x_mean, y = y_mean),
              color = "blue", size = 3, alpha = 0.6) +
    geom_point(data = binned_right, aes(x = x_mean, y = y_mean),
              color = "red", size = 3, alpha = 0.6) +
    # Fitted lines
    geom_line(data = data.frame(x = x_plot_left, y = y_plot_left),
             aes(x = x, y = y), color = "blue", linewidth = 1.5) +
    geom_line(data = data.frame(x = x_plot_right, y = y_plot_right),
             aes(x = x, y = y), color = "red", linewidth = 1.5) +
    # Cutoff line
    geom_vline(xintercept = cutoff, linetype = "dashed",
              linewidth = 1, alpha = 0.7) +
    labs(
      x = "Running Variable (centered at cutoff)",
      y = "Outcome",
      title = sprintf("RDD Plot (Bandwidth = %.3f)", h),
      subtitle = sprintf("Treatment Effect = %.3f (SE = %.3f)",
                        rdd_result$coef[1], rdd_result$se[1])
    ) +
    theme_minimal() +
    theme(
      text = element_text(size = 12),
      plot.title = element_text(size = 14, face = "bold"),
      plot.subtitle = element_text(size = 11)
    )

  print(p)
  invisible(p)
}

# ============================================
# Bandwidth Sensitivity Analysis
# ============================================

bandwidth_sensitivity <- function(data, running_var = "running_var",
                                  outcome = "outcome", cutoff = 0,
                                  h_range = NULL, n_points = 20) {
  #' Test sensitivity to bandwidth choice
  #'
  #' @param data data.frame
  #' @param running_var Running variable name
  #' @param outcome Outcome variable name
  #' @param cutoff Cutoff value
  #' @param h_range Vector of min and max bandwidth
  #' @param n_points Number of bandwidth values to try

  x <- data[[running_var]]
  y <- data[[outcome]]

  # Default bandwidth range
  if (is.null(h_range)) {
    h_range <- c(0.3, 2.0)
  }

  bandwidths <- seq(h_range[1], h_range[2], length.out = n_points)

  results <- data.frame(
    bandwidth = numeric(n_points),
    estimate = numeric(n_points),
    se = numeric(n_points),
    ci_lower = numeric(n_points),
    ci_upper = numeric(n_points)
  )

  for (i in seq_along(bandwidths)) {
    h <- bandwidths[i]
    rdd <- rdrobust(y, x, c = cutoff, h = h)

    results$bandwidth[i] <- h
    results$estimate[i] <- rdd$coef[1]
    results$se[i] <- rdd$se[1]
    results$ci_lower[i] <- rdd$ci[1, 1]
    results$ci_upper[i] <- rdd$ci[1, 2]
  }

  # Plot sensitivity
  p <- ggplot(results, aes(x = bandwidth, y = estimate)) +
    geom_line(linewidth = 1) +
    geom_point(size = 2) +
    geom_ribbon(aes(ymin = ci_lower, ymax = ci_upper),
               alpha = 0.3, fill = "blue") +
    geom_hline(yintercept = 5.0, linetype = "dashed",
              color = "red", linewidth = 1) +
    annotate("text", x = mean(h_range), y = 5.0,
            label = "True Effect", vjust = -0.5, color = "red") +
    labs(
      x = "Bandwidth",
      y = "Estimated Treatment Effect",
      title = "RDD Sensitivity to Bandwidth Choice",
      subtitle = "Shaded area shows 95% confidence interval"
    ) +
    theme_minimal() +
    theme(text = element_text(size = 12))

  print(p)
  invisible(results)
}

# ============================================
# Density Continuity Test (Manipulation Check)
# ============================================

test_manipulation <- function(data, running_var = "running_var", cutoff = 0) {
  #' Test for manipulation of running variable (McCrary density test)
  #'
  #' @param data data.frame
  #' @param running_var Running variable name
  #' @param cutoff Cutoff value

  x <- data[[running_var]]

  # Density test
  density_test <- rddensity(x, c = cutoff)

  cat("\n")
  cat("=" * 70, "\n")
  cat("Manipulation Test (Density Continuity)\n")
  cat("=" * 70, "\n")
  print(summary(density_test))
  cat("\n")

  # Plot density
  p <- rdplotdensity(density_test, x)

  invisible(density_test)
}

# ============================================
# Example 1: Remedial Education Program
# ============================================

example_education_rdd <- function() {
  cat("Example 1: Remedial Education Program\n")
  cat("======================================\n\n")

  set.seed(456)
  n <- 2000

  # Test score (running variable)
  test_score <- rnorm(n, 50, 15)

  # Treatment: remedial program if score < 50
  treatment <- as.integer(test_score < 50)

  # Outcome: next year's test score
  x_centered <- test_score - 50
  true_effect <- 8.0
  next_score <- 55 + 0.5*x_centered + 0.02*x_centered^2 +
               true_effect * treatment + rnorm(n, 0, 5)

  df <- data.frame(
    test_score = test_score,
    next_score = next_score,
    treatment = treatment
  )

  # Estimate RDD
  cat("Estimating RDD with MSE-optimal bandwidth...\n\n")
  rdd <- rdrobust(df$next_score, df$test_score, c = 50)

  # Print results
  print(summary(rdd))

  cat("\n")
  cat(sprintf("True effect: %.2f\n", true_effect))
  cat(sprintf("Estimated effect: %.3f (SE = %.3f)\n",
             rdd$coef[1], rdd$se[1]))
  cat(sprintf("95%% CI: [%.3f, %.3f]\n", rdd$ci[1,1], rdd$ci[1,2]))
  cat(sprintf("Bandwidth: %.3f\n", rdd$bws[1]))
  cat(sprintf("N (left): %d, N (right): %d\n",
             rdd$N_h[1], rdd$N_h[2]))

  # Plot
  df$running_var <- df$test_score
  df$outcome <- df$next_score
  p <- plot_rdd(df, rdd, cutoff = 50)
  ggsave("rdd_education_R.png", plot = p, width = 10, height = 6, dpi = 150)

  # Manipulation test
  test_manipulation(df, cutoff = 50)

  invisible(rdd)
}

# ============================================
# Example 2: Bandwidth Sensitivity
# ============================================

example_bandwidth_sensitivity <- function() {
  cat("\nExample 2: Bandwidth Sensitivity Analysis\n")
  cat("==========================================\n\n")

  # Simulate data
  df <- simulate_rdd_data(n = 1500, effect = 5.0)

  # Sensitivity analysis
  sensitivity_results <- bandwidth_sensitivity(df)
  ggsave("rdd_bandwidth_sensitivity_R.png", width = 10, height = 6, dpi = 150)

  invisible(sensitivity_results)
}

# ============================================
# Example 3: Fuzzy RDD
# ============================================

example_fuzzy_rdd <- function() {
  cat("\nExample 3: Fuzzy RDD (Imperfect Compliance)\n")
  cat("===========================================\n\n")

  set.seed(789)
  n <- 1500

  # Running variable
  x <- runif(n, -2, 2)

  # Treatment eligibility
  eligible <- as.integer(x >= 0)

  # Actual treatment (fuzzy: not everyone complies)
  # Compliance rate: 80% above cutoff, 20% below
  compliance_prob <- ifelse(eligible == 1, 0.8, 0.2)
  treatment <- rbinom(n, 1, compliance_prob)

  # Outcome
  true_effect <- 6.0
  y <- 10 + 2*x + 0.5*x^2 + true_effect * treatment + rnorm(n, 0, 1)

  # Fuzzy RDD (uses eligibility as instrument for treatment)
  fuzzy_rdd <- rdrobust(y, x, c = 0, fuzzy = treatment)

  cat("Fuzzy RDD Results:\n")
  print(summary(fuzzy_rdd))

  cat("\n")
  cat(sprintf("True effect: %.2f\n", true_effect))
  cat(sprintf("Estimated LATE: %.3f (SE = %.3f)\n",
             fuzzy_rdd$coef[1], fuzzy_rdd$se[1]))

  # First stage (effect on treatment uptake)
  first_stage <- rdrobust(treatment, x, c = 0)
  cat(sprintf("\nFirst stage (jump in treatment): %.3f\n",
             first_stage$coef[1]))

  invisible(fuzzy_rdd)
}

# ============================================
# Run all examples
# ============================================

if (sys.nframe() == 0) {
  cat("\n")
  cat("=" * 70, "\n")
  cat("Regression Discontinuity Design Examples in R\n")
  cat("=" * 70, "\n\n")

  example_education_rdd()
  example_bandwidth_sensitivity()
  example_fuzzy_rdd()

  cat("\nAll examples completed successfully!\n")
}

File: 04-causal-inference/causal-inference-fundamentals/code/python/regression_discontinuity.py
"""
Regression Discontinuity Design (RDD)
======================================

Implementation of sharp and fuzzy RDD with various estimation methods.

References:
    Imbens & Lemieux (2008), Lee & Lemieux (2010), Calonico et al. (2014)
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from typing import Tuple, Optional
from dataclasses import dataclass

sns.set_style("whitegrid")


@dataclass
class RDDResults:
    """Results from RDD estimation."""
    treatment_effect: float
    standard_error: float
    ci_lower: float
    ci_upper: float
    bandwidth: float
    n_left: int
    n_right: int


class RegressionDiscontinuity:
    """
    Regression Discontinuity Design estimator.

    Parameters
    ----------
    running_var : array-like
        Running variable (forcing variable)
    outcome : array-like
        Outcome variable
    treatment : array-like, optional
        Treatment indicator (for fuzzy RDD)
    cutoff : float
        Cutoff value for treatment assignment
    """

    def __init__(self, running_var: np.ndarray, outcome: np.ndarray,
                 treatment: Optional[np.ndarray] = None, cutoff: float = 0.0):
        self.running_var = np.asarray(running_var)
        self.outcome = np.asarray(outcome)
        self.treatment = treatment if treatment is None else np.asarray(treatment)
        self.cutoff = cutoff
        self.n = len(running_var)

        # Center running variable at cutoff
        self.x_centered = self.running_var - self.cutoff

    def imbens_kalyanaraman_bandwidth(self) -> float:
        """
        Calculate optimal bandwidth using Imbens-Kalyanaraman method.

        Returns
        -------
        float
            Optimal bandwidth
        """
        # Pilot bandwidth using rule of thumb
        h_pilot = 1.84 * np.std(self.x_centered) * self.n ** (-1/5)

        # Estimate derivatives on both sides
        mask_left = (self.x_centered < 0) & (abs(self.x_centered) < h_pilot)
        mask_right = (self.x_centered >= 0) & (abs(self.x_centered) < h_pilot)

        # Fit cubic polynomials
        poly = PolynomialFeatures(degree=3)

        # Left side
        if mask_left.sum() > 10:
            X_left = poly.fit_transform(self.x_centered[mask_left].reshape(-1, 1))
            y_left = self.outcome[mask_left]
            model_left = LinearRegression().fit(X_left, y_left)
            m2_left = 2 * model_left.coef_[2]  # Second derivative
        else:
            m2_left = 0

        # Right side
        if mask_right.sum() > 10:
            X_right = poly.fit_transform(self.x_centered[mask_right].reshape(-1, 1))
            y_right = self.outcome[mask_right]
            model_right = LinearRegression().fit(X_right, y_right)
            m2_right = 2 * model_right.coef_[2]
        else:
            m2_right = 0

        # Estimate variance
        sigma2_left = np.var(y_left) if mask_left.sum() > 0 else 1
        sigma2_right = np.var(y_right) if mask_right.sum() > 0 else 1

        # IK bandwidth formula
        N = self.n
        C_k = 3.4375  # Constant for rectangular kernel
        m2 = abs(m2_left) + abs(m2_right)

        if m2 > 0:
            h_ik = C_k * (sigma2_left + sigma2_right) / (N * m2**2)
            h_ik = h_ik ** (1/5)
        else:
            h_ik = h_pilot

        return h_ik

    def local_linear_regression(self, bandwidth: float, polynomial_order: int = 1) -> RDDResults:
        """
        Estimate treatment effect using local linear regression.

        Parameters
        ----------
        bandwidth : float
            Bandwidth for local regression
        polynomial_order : int
            Order of polynomial (1 for local linear, 2 for local quadratic)

        Returns
        -------
        RDDResults
            Estimation results
        """
        # Select observations within bandwidth
        mask = abs(self.x_centered) <= bandwidth
        x_local = self.x_centered[mask]
        y_local = self.outcome[mask]

        # Create treatment indicator
        treated = (x_local >= 0).astype(float)

        # Create polynomial features
        poly = PolynomialFeatures(degree=polynomial_order, include_bias=False)
        x_poly = poly.fit_transform(x_local.reshape(-1, 1))

        # Interaction terms: treatment * polynomial features
        X = np.column_stack([
            np.ones(len(x_local)),  # Intercept
            treated,  # Treatment indicator
            x_poly,  # Polynomial of running variable
            treated.reshape(-1, 1) * x_poly  # Interactions
        ])

        # Triangular kernel weights
        weights = (1 - abs(x_local / bandwidth)) * mask[mask]

        # Weighted least squares
        W = np.diag(weights)
        X_weighted = np.sqrt(W) @ X
        y_weighted = np.sqrt(W) @ y_local

        # Estimate
        beta = np.linalg.lstsq(X_weighted, y_weighted, rcond=None)[0]
        treatment_effect = beta[1]

        # Standard error (heteroskedasticity-robust)
        residuals = y_local - X @ beta
        meat = X.T @ (W * (residuals**2)[:, np.newaxis] * X)
        bread_inv = np.linalg.inv(X.T @ W @ X)
        var_beta = bread_inv @ meat @ bread_inv
        se = np.sqrt(var_beta[1, 1])

        # Confidence interval
        ci_lower = treatment_effect - 1.96 * se
        ci_upper = treatment_effect + 1.96 * se

        n_left = (x_local < 0).sum()
        n_right = (x_local >= 0).sum()

        return RDDResults(
            treatment_effect=treatment_effect,
            standard_error=se,
            ci_lower=ci_lower,
            ci_upper=ci_upper,
            bandwidth=bandwidth,
            n_left=n_left,
            n_right=n_right
        )

    def estimate(self, bandwidth: Optional[float] = None,
                 polynomial_order: int = 1) -> RDDResults:
        """
        Estimate RDD effect.

        Parameters
        ----------
        bandwidth : float, optional
            Bandwidth (if None, uses Imbens-Kalyanaraman)
        polynomial_order : int
            Polynomial order for local regression

        Returns
        -------
        RDDResults
            Estimation results
        """
        if bandwidth is None:
            bandwidth = self.imbens_kalyanaraman_bandwidth()
            print(f"Using IK bandwidth: {bandwidth:.3f}")

        return self.local_linear_regression(bandwidth, polynomial_order)

    def plot_rdd(self, results: RDDResults, bins: int = 20):
        """
        Create RDD visualization.

        Parameters
        ----------
        results : RDDResults
            Estimation results
        bins : int
            Number of bins for binned scatter plot
        """
        fig, axes = plt.subplots(1, 2, figsize=(14, 5))

        # Left panel: Binned scatter plot
        ax = axes[0]

        # Create bins
        bins_left = pd.cut(self.x_centered[self.x_centered < 0], bins=bins//2)
        bins_right = pd.cut(self.x_centered[self.x_centered >= 0], bins=bins//2)

        # Calculate bin means
        df_left = pd.DataFrame({'x': self.x_centered[self.x_centered < 0],
                               'y': self.outcome[self.x_centered < 0],
                               'bin': bins_left})
        df_right = pd.DataFrame({'x': self.x_centered[self.x_centered >= 0],
                                'y': self.outcome[self.x_centered >= 0],
                                'bin': bins_right})

        bin_means_left = df_left.groupby('bin')[['x', 'y']].mean()
        bin_means_right = df_right.groupby('bin')[['x', 'y']].mean()

        # Plot binned means
        ax.scatter(bin_means_left['x'], bin_means_left['y'],
                  color='blue', s=50, alpha=0.6, label='Control')
        ax.scatter(bin_means_right['x'], bin_means_right['y'],
                  color='red', s=50, alpha=0.6, label='Treatment')

        # Fit lines within bandwidth
        mask_left = (self.x_centered < 0) & (abs(self.x_centered) <= results.bandwidth)
        mask_right = (self.x_centered >= 0) & (abs(self.x_centered) <= results.bandwidth)

        # Left polynomial
        if mask_left.sum() > 0:
            x_plot_left = np.linspace(self.x_centered[mask_left].min(), 0, 100)
            poly_left = np.polyfit(self.x_centered[mask_left], self.outcome[mask_left], 1)
            y_plot_left = np.polyval(poly_left, x_plot_left)
            ax.plot(x_plot_left, y_plot_left, 'b-', linewidth=2, label='Fit (Control)')

        # Right polynomial
        if mask_right.sum() > 0:
            x_plot_right = np.linspace(0, self.x_centered[mask_right].max(), 100)
            poly_right = np.polyfit(self.x_centered[mask_right], self.outcome[mask_right], 1)
            y_plot_right = np.polyval(poly_right, x_plot_right)
            ax.plot(x_plot_right, y_plot_right, 'r-', linewidth=2, label='Fit (Treatment)')

        # Vertical line at cutoff
        ax.axvline(x=0, color='black', linestyle='--', linewidth=1.5, alpha=0.7)

        ax.set_xlabel('Running Variable (centered at cutoff)')
        ax.set_ylabel('Outcome')
        ax.set_title(f'RDD Plot (Bandwidth = {results.bandwidth:.3f})')
        ax.legend()

        # Right panel: Density of running variable (continuity check)
        ax = axes[1]
        ax.hist(self.x_centered[self.x_centered < 0], bins=30, alpha=0.5,
               label='Control', density=True, color='blue')
        ax.hist(self.x_centered[self.x_centered >= 0], bins=30, alpha=0.5,
               label='Treatment', density=True, color='red')
        ax.axvline(x=0, color='black', linestyle='--', linewidth=1.5, alpha=0.7)
        ax.set_xlabel('Running Variable (centered at cutoff)')
        ax.set_ylabel('Density')
        ax.set_title('Density of Running Variable\n(Continuity Check)')
        ax.legend()

        plt.tight_layout()
        return fig


def simulate_rdd_data(n: int = 1000, effect: float = 5.0,
                     polynomial_order: int = 2) -> pd.DataFrame:
    """
    Simulate RDD data.

    Parameters
    ----------
    n : int
        Sample size
    effect : float
        True treatment effect
    polynomial_order : int
        Polynomial order for baseline relationship

    Returns
    -------
    pd.DataFrame
        Simulated data
    """
    np.random.seed(42)

    # Running variable
    x = np.random.uniform(-2, 2, n)

    # Treatment assignment (sharp RDD)
    D = (x >= 0).astype(float)

    # Outcome with polynomial trend
    y_baseline = 10 + 2*x + 0.5*x**2
    if polynomial_order >= 3:
        y_baseline += 0.1*x**3

    # Add treatment effect
    y = y_baseline + effect * D + np.random.normal(0, 1, n)

    return pd.DataFrame({'x': x, 'y': y, 'D': D})


def example_education_rdd():
    """
    Example: Effect of remedial education program.

    Students below cutoff score are assigned to remedial program.
    """
    print("Example: Remedial Education Program RDD")
    print("=" * 70)

    # Simulate data
    np.random.seed(456)
    n = 2000

    # Test score (running variable, centered at cutoff of 50)
    test_score = np.random.normal(50, 15, n)
    x_centered = test_score - 50

    # Treatment: remedial program if score < 50
    treatment = (test_score < 50).astype(float)

    # Outcome: next year's test score
    # Baseline relationship + treatment effect
    true_effect = 8.0
    next_score = 55 + 0.5*x_centered + 0.02*x_centered**2 + \
                 true_effect * treatment + np.random.normal(0, 5, n)

    # Estimate RDD
    rdd = RegressionDiscontinuity(running_var=test_score, outcome=next_score, cutoff=50)
    results = rdd.estimate()

    print(f"Estimated treatment effect: {results.treatment_effect:.3f}")
    print(f"Standard error: {results.standard_error:.3f}")
    print(f"95% CI: [{results.ci_lower:.3f}, {results.ci_upper:.3f}]")
    print(f"True effect: {true_effect:.3f}")
    print(f"Bandwidth: {results.bandwidth:.3f}")
    print(f"N (control): {results.n_left}")
    print(f"N (treatment): {results.n_right}")

    # Plot
    fig = rdd.plot_rdd(results)
    plt.savefig('rdd_education_example.png', dpi=150, bbox_inches='tight')
    plt.show()


def example_bandwidth_sensitivity():
    """Test sensitivity to bandwidth choice."""
    print("\n" + "=" * 70)
    print("Bandwidth Sensitivity Analysis")
    print("=" * 70)

    # Simulate data
    df = simulate_rdd_data(n=1500, effect=5.0)

    rdd = RegressionDiscontinuity(running_var=df['x'].values,
                                  outcome=df['y'].values,
                                  cutoff=0)

    # Try different bandwidths
    bandwidths = np.linspace(0.3, 2.0, 20)
    effects = []
    ci_lower = []
    ci_upper = []

    for bw in bandwidths:
        results = rdd.local_linear_regression(bandwidth=bw)
        effects.append(results.treatment_effect)
        ci_lower.append(results.ci_lower)
        ci_upper.append(results.ci_upper)

    # Plot sensitivity
    fig, ax = plt.subplots(figsize=(10, 6))
    ax.plot(bandwidths, effects, 'o-', linewidth=2, markersize=6, label='Point estimate')
    ax.fill_between(bandwidths, ci_lower, ci_upper, alpha=0.3, label='95% CI')
    ax.axhline(y=5.0, color='red', linestyle='--', linewidth=2, label='True effect')
    ax.set_xlabel('Bandwidth')
    ax.set_ylabel('Estimated Treatment Effect')
    ax.set_title('RDD Sensitivity to Bandwidth Choice')
    ax.legend()
    ax.grid(True, alpha=0.3)

    plt.tight_layout()
    plt.savefig('rdd_bandwidth_sensitivity.png', dpi=150, bbox_inches='tight')
    plt.show()


if __name__ == "__main__":
    example_education_rdd()
    example_bandwidth_sensitivity()
